### DISCUSSION TITLE
How do I run the SIONNA 5G-LDPC codes for BPSK modulation?

### DISCUSSION THREAD

**Discussion by yossefyossef** [NVlabs/sionna#109] [Q&A]
Hi all,

Can someone please help me, how to run the SIONNA for BPSK modulation, for the long codes of LDPC?

That means, I write the following value:
num_bits_per_symbol = 1  # BPSK

I got the following error message: 
“num_bits_per_symbol must be multiple of 2”.

Can someone please help me?

Thanking in advance, and best regards,
Yossef Yossef
Email: [yossiy@gilat.com](mailto:yossiy@gilat.com)

**Comment by yossefyossef** [comment]
Note:
I use the code from the following link:
https://nvlabs.github.io/sionna/examples/5G_Channel_Coding_Polar_vs_LDPC_Codes.html

**Comment by jhoydis** [comment]
Hi Yossef,

For BPSK, you need to use PAM modulation. See https://nvlabs.github.io/sionna/api/mapping.html#sionna.mapping.Mapper
This code would give you a BPSK modulation:
```
mapper = Mapper("pam", 1)
```
Hope this helps.

**Reply by yossefyossef** [reply]
Dear Jacob,
Thank you very much for your help!

I would like to be sure that I understand your answer:

I should write the following two rows:

mapper = Mapper("pam", 1)  #this is the added row
num_bits_per_symbol = 1  # BPSK

Do I understand correctly?

**Reply by jhoydis** [reply]
No, ``num_bits_per_symbol`` is already set to 1 in the first line. Here is the code spelling the keyword arguments out:

```
mapper = Mapper(constellation_type="pam", num_bits_per_symbol=1)
```

The ``mapper`` maps now a bit stream to BPSK symbols.

**Reply by yossefyossef** [reply]
So, instead to write:
num_bits_per_symbol = 1 # BPSK

I need to write the row:
mapper = Mapper("pam", 1) 

Am I correct?

**Reply by jhoydis** [reply]
Yes. This is all plain Python code. When you write ``num_bits_per_symbol = 1``, it simply defines a variable named `num_bits_per_symbol ` which is equal to one. If you do not use this variable anywhere, it has no impact.

So, you could also write
```
num_bits_per_symbol = 1
mapper = Mapper("pam", num_bits_per_symbol)
```

**Reply by yossefyossef** [reply]
Jacob, you are an excellent teacher!

**Reply by yossefyossef** [reply]
I did it, and now I got a new error message:

ValueError: Unsupported coderate (r<1/5).

I used the following parameters:
k= 1920
n= 16384
Code-rate R= 0.1171875

Do you have a solution for this case?

**Reply by SebastianCa** [reply]
This is related to discussion #110. Please ensure that the coderate r>1/3. In this case, you could reduce the code length n (e.g., by a factor of 4).

**Comment by yossefyossef** [comment]
I made the following attempts to run the LDPC for BPSK, and they gave an error messages.

Maybe you can help me?

Attempt #1:
---------------

k= 1920
n = 4800
Code-rate R= 0.4

Used the following row:
mapper = Mapper("pam", 1) #this is the added row

got the error message:
NameError: name 'num_bits_per_symbol' is not defined


Attempt #2:
---------------

k= 1920
n = 4800
Code-rate R= 0.4

Used the following rows:
num_bits_per_symbol = 1
mapper = Mapper("pam", num_bits_per_symbol)

got the error message:
AssertionError: num_bits_per_symbol must be a mutliple of 2

Can you please guide me how to run the simulation for BPSK? It seems that I am missing out something.

**Reply by jhoydis** [reply]
Hi,

Which version of Sionna are you using?
This line definitely works:
```
Mapper("pam", 1)
```

**Reply by yossefyossef** [reply]
I wrote this row as it:
Mapper("pam", 1)
And got the error:
NameError: name 'num_bits_per_symbol' is not defined

I am checking which Sionna version was installed..

**Comment by jhoydis** [comment]
Please do this in a clean notebook:
```
import sionna
from sionna.mapping import Mapper
print(sionna.__version__)
Mapper("pam", 1)
```

**Reply by yossefyossef** [reply]
I use Sionna version 0.11.0

**Comment by yossefyossef** [comment]
I run the suggested rows:
import sionna
from sionna.mapping import Mapper
print(sionna.__version__)
Mapper("pam", 1)

and got the following result:

C:\Anaconda3\envs\tf\python.exe C:\Users\modem\PycharmProjects\sionna_test1\test_Sionna_ver.py 
0.11.0
2023-04-03 11:06:45.765661: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX AVX2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-04-03 11:06:46.541939: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1616] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 6647 MB memory:  -> device: 0, name: Quadro P4000, pci bus id: 0000:01:00.0, compute capability: 6.1

Process finished with exit code 0

**Comment by jhoydis** [comment]
It should work with Sionna 0.11. You can try upgrading to a later version.
I cannot reproduce your issue.
